Add support bundle docs (#4457)
antrea-io/antrea: docs/support-bundle-guide.md

> # Support Bundle User Guide

Reviewer: I suppose we need link this file to somewhere to let user know about this new user guide?
If there is no proper place, you may add a line the `Collecting support information` section to let user know about other way to collect support bundle.

> ## What is Supportbundle

Reviewer: ```suggestion
## What is Support Bundle
```

> Antrea supports creating support bundle tarballs, which include information from the cluster,

Reviewer: Done

the Antrea Controller, Antrea Agents and ExternalNode managed by the cluster.
The information is useful when troubleshooting issues in Kubernetes clusters using Antrea.

**Be aware that the generated support bundle includes a lot of information,
  including logs, so please review the contents of the directory before sharing
  it on Github and ensure that you do not share anything sensitive.**

There are two ways of generating support bundles. Firstly, You can run `antctl supportbundle`

> directly in the Antrea Agent Pod, Antrea Controller Pod, or on the Host

Reviewer: ```suggestion
directly in the Antrea Agent Pod, Antrea Controller Pod, or on a host
```

> with `kubeconfig` targeting Antrea Cluster. Secondly, You can also apply

Reviewer: ```suggestion
with `kubeconfig` targeting Antrea cluster. Secondly, You can also apply
```

> with `kubeconfig` targeting Antrea Cluster. Secondly, You can also apply

Reviewer 2: ```suggestion
with `kubeconfig` targeting the cluster. Secondly, You can also apply
```

> SupportBundleCollection YAMLs to create support bundles for Kubernetes Nodes or external Nodes.

Reviewer 2: ```suggestion
SupportBundleCollection CRs to create support bundles for Kubernetes Nodes or external Nodes.
```

> We use "SupportBundle Collection" to designate such feature. The details will be provided

Reviewer: SupportBundle Collection -> SupportBundleCollection

> We use "SupportBundle Collection" to designate such feature. The details will be provided

Reviewer 2: ```suggestion
We use "SupportBundle Collection" to designate this feature. The details will be provided
```

in section [Use case example](#use-case-example).

## Table of Contents

<!-- toc -->
- [Prerequisites](#prerequisites)
- [Use case example](#use-case-example)
  - [Running antctl commands](#running-antctl-commands)
  - [Applying SupportBundleCollection CRD](#applying-supportbundlecollection-crd)
- [List of collected items](#list-of-collected-items)
- [Limitations](#limitations)
<!-- /toc -->

## Prerequisites

> SupportBundle Collection is introduced in Antrea v1.10.0 as an alpha feature. The feature gate

Reviewer: SupportBundle Collection -> SupportBundleCollection

> `SupportBundleCollection` must be enabled in the `antrea-agent` configuration to use the feature.

Reviewer 2: The feature gate must be enabled in both antrea-controller and antrea-agent configurations

Reviewer 2: Remove "to use the feature"?

> If you are planning to collect support bundle on the specified ExternalNode, you should enable

Reviewer: ```suggestion
If you are planning to collect support bundle on a specific ExternalNode, you should enable
```

> If you are planning to collect support bundle on the specified ExternalNode, you should enable

Reviewer 2: ```suggestion
If you are planning to collect support bundle on external Nodes, you should enable
```

> `SupportBundleCollection` on that ExternalNode as well.

Reviewer: Is it the same way to enable `SupportBundleCollection` for ExternalNode?

> `SupportBundleCollection` on that ExternalNode as well.

Reviewer 2: ```suggestion
it in the configuration on those ExternalNode as well.
```

```yaml
  antrea-agent.conf: |
    featureGates:
    # Enable collecting support bundle files with SupportBundleCollection CRD.
    SupportBundleCollection: true

> ```

Reviewer 2: Please also provide the sample yaml change in antrea-controller.conf.

## Use case example

### Running antctl commands

> Please refer to the corresponding [antctl user guide section](antctl.md#collecting-support-information).

Reviewer 2: Add description about the limitation that "antctl supportbundle" only works on collecting the bundles from Controller and Agent which is working on K8s Node, but does not work for the Agent on an ExternalNode.

### Applying SupportBundleCollection CRD

> In this section, we will create two SupportBundleCollections for Antrea Nodes and External Node.

Reviewer: ```suggestion
In this section, we will create two SupportBundleCollections for Antrea Nodes and ExternalNodes.
```

> In this section, we will create two SupportBundleCollections for Antrea Nodes and External Node.

Reviewer 2: ```suggestion
In this section, we will create two SupportBundleCollections for Kubernetes Nodes and External Node.
```

Assume we have a cluster with Nodes named "kube-worker" and "kube-master". In addition, we have

> set up an ExternalNode named `vm` in `vm-ns` namespace by following the instruction

Reviewer: namespace -> Namespace

> set up an ExternalNode named `vm` in `vm-ns` namespace by following the instruction

Reviewer 2: ```suggestion
set up an ExternalNode named as `vm` in the `vm-ns` namespace by following the instruction
```

in [Install ExternalNode on VM](external-node.md#install-antrea-agent-on-vm).

> A Secret need to be created first with the username and password of the SSH Server.

Reviewer 2: ```suggestion
A Secret needs to be created in advance with the username and password of the SFTP Server.
```

> The Secret will be used in `authSecret` in the following YAML examples.

Reviewer 2: Add a note that the username and password values are encoded with base64 in the Secret ?

```bash

> kubectl create secret generic support-bundle-secret  --from-literal=username='yoursshusername'  --from-literal=password='yoursshpassword'

Reviewer: ```suggestion
kubectl create secret generic support-bundle-secret  --from-literal=username='your-ssh-username'  --from-literal=password='your-ssh-password'
```

> kubectl create secret generic support-bundle-secret  --from-literal=username='yoursshusername'  --from-literal=password='yoursshpassword'

Reviewer 2: ```suggestion
kubectl create secret generic support-bundle-secret  --from-literal=username='yoursftpusername'  --from-literal=password='yoursftppassword'
```

```

Then we could apply the following YAML files.
The first one is to collect support bundle on Antrea Nodes "kube-worker" and "kube-master".
The second one is to collect support bundle on ExternalNode "vm".

```yaml
kind: SupportBundleCollection
metadata:
  name: collect-support-bundle-for-kube-nodes
spec:
  nodes:

>     nodeNames: // empty list means all the Nodes

Reviewer: ```suggestion
    nodeNames: // empty list means all Nodes
```

      - kube-master
      - kube-worker
  expirationMinutes: 10 // expirationMinutes is the requested duration of validity of the SupportBundleCollection. A SupportBundleCollection will be marked as Failed if it does not finish before expiration.
  fileServer:
    url: sftp://yourtestdomain.com:22/root/test 
  authentication:
    authType: "BasicAuthentication"
    authSecret:
      name: support-bundle-secret
      namespace: default
```

```yaml
apiVersion: crd.antrea.io/v1alpha1
kind: SupportBundleCollection
metadata:
  name: collect-support-bundle-for-vm-external-node
spec:
  externalNodes:

>     nodeNames: // empty list means all the ExternalNodes in namespace

Reviewer: ```suggestion
    nodeNames: # empty list means all ExternalNodes in a specific Namespace
```

`//` It's not a comment signature in yaml, please correct them.

>     nodeNames: // empty list means all the ExternalNodes in namespace

Reviewer 2: ```suggestion
    nodeNames: // empty list means all the ExternalNodes in the namespace
```

>       - vm

Reviewer 2: I think you should also give an introduction that we support using `                    nodeSelector` to select Nodes/ExternalNodes by labels?

Reviewer: Also need your help for this comment, could you provide a proper description and sample for this? thanks.

Reviewer 2: We can add this sample in "externalNodes" block,

```
    nodeNames: # Empty list means all the ExternalNodes in the Namespace
      - vm 
    nodeSelector:
       matchLabels:
         app: db   # Collect support bundle files from the ExternalNodes which are configured with label "app: db" in the Namespace
    namespace: vm-ns
```

Author: updated. Thanks

    namespace: vm-ns
  fileServer:
    url: yourtestdomain.com:22/root/test //scheme sftp can be omitted
  authentication:
    authType: "BasicAuthentication"
    authSecret:
      name: support-bundle-secret
      namespace: default
```

You can check the status of SupportBundleCollection by running
`kubectl get supportbundlecollections NameOfSupportBundleCollection -ojson`. The following example

> shows a successful creation of a SupportBundleCollection. `collectedNodes` and `desiredNodes`

Reviewer 2: `desiredNodes` shows the number of Nodes/ExternalNodes is expected in this request, while `collectedNodes` shows the number of Nodes/ExternalNodes which have already uploaded bundle files to the target file server.

> should match the number of Nodes/ExternalNodes you want to collect support bundle. There is also

Reviewer 2: These two conditions are presented when a bundle creation is succeeded, 
1. "Completed" is true
2. "CollectionFailure" is false.

If any expected Node/ExternalNode is failed to upload the bundle files in the required time, the "CollectionFailure" condition is marked as true.

an entry of "Completed" status in the "conditions" list.

```bash
$ kubectl get supportbundlecollections collect-support-bundle-for-kube-nodes -ojson

...
 "status": {
        "collectedNodes": 1,
        "conditions": [
            {
                "lastTransitionTime": "2022-12-08T06:49:35Z",
                "status": "True",
                "type": "Started"
            },
            {
                "lastTransitionTime": "2022-12-08T06:49:41Z",
                "status": "True",
                "type": "BundleCollected"
            },
            {
                "lastTransitionTime": "2022-12-08T06:49:35Z",
                "status": "False",
                "type": "CollectionFailure"
            },
            {
                "lastTransitionTime": "2022-12-08T06:49:41Z",
                "status": "True",
                "type": "Completed"
            }
        ],
        "desiredNodes": 1
    }
```

> To access the files, SSH into yourtestdomain.com with port 22. Three tarballs,

Reviewer 2: To access the files, SSH into yourtestdomain.com with port 22, or scp them to any other storages you expected.

> To access the files, SSH into yourtestdomain.com with port 22. Three tarballs,
> two for `collect-support-bundle-for-kube-nodes` and one for
> `collect-support-bundle-for-vm-external-node` are expected to be shown in `/root/test` folder.
> Run the `tar xvf TARBALL_NAME` command and you will get the files,
> each with different collect items.

Reviewer: ```suggestion
The output contains three tarballs. To access these files, please SSH into the ftp server `yourtestdomain.com` with port 22. There will be two tarballs for `collect-support-bundle-for-kube-nodes` and one for
`collect-support-bundle-for-vm-external-node` showing in `/root/test` folder.
Run the `tar xvf TARBALL_NAME` command and you will get extracted files.
Each tarball contains different collect items:
```

I am not sure I got the idea about ` two tarballs for `collect-support-bundle-for-kube-nodes` `, I suppose you mean there will be one tarball for a K8s Node, and you have two Nodes, so there are two tarballs?
I feel it's better to show the output of `/root/test` to let user understand the output.

Reviewer: @wenyingd do you have this kind of information?

> each with different collect items.

Reviewer 2: Add a simple description about eventual file name uploaded from Agent.

## List of collected items

Depending on the methods you use to collect support bundle, the items you collect may differ.
Here is the table that shows such differences.

We use `A`,`C` and `O` to represent "running `antctl supportbundle` in the Antrea Agent",
"running `antctl supportbundle` in the Antrea Controller" and
"running `antctl supportbundle` out-of-cluster". Also, We use `SBC-N` and `SBC-E` to represent
"apply SupportBundleCollection CRD to create support bundle on Nodes" and
"apply SupportBundleCollection CRD to create support bundle on ExternalNodes".

> Please refer to the corresponding [antctl user guide section](antctl.md) for the detail of `antctl` commands.

Reviewer: I don't get this, should this be removed?

Author: I mean there are lots of antctl commands in the following table, which may confuse the readers. He/She can check the antctl document for detail

> | Collected Item                | Supported Collecting Method                     | Explanation       |

Reviewer: Please format the table.

"CollectionFailure" condition is marked as true.

```bash
$ kubectl get supportbundlecollections collect-support-bundle-for-kube-nodes -ojson

...
 "status": {
        "collectedNodes": 1,
        "conditions": [
            {
                "lastTransitionTime": "2022-12-08T06:49:35Z",
                "status": "True",
                "type": "Started"
            },
            {
                "lastTransitionTime": "2022-12-08T06:49:41Z",
                "status": "True",
                "type": "BundleCollected"
            },
            {
                "lastTransitionTime": "2022-12-08T06:49:35Z",
                "status": "False",
                "type": "CollectionFailure"
            },
            {
                "lastTransitionTime": "2022-12-08T06:49:41Z",
                "status": "True",
                "type": "Completed"
            }
        ],
        "desiredNodes": 1
    }
```

The output contains three tarballs. To access these files, please SSH into the ftp server
`yourtestdomain.com` with port 22 or scp them to any other storages you used in CR. There will be

> two tarballs for `collect-support-bundle-for-kube-nodes` and one for `collect-support-bundle-for-vm-external-node`

Reviewer 2: two tarballs for `collect-support-bundle-for-kube-nodes` ( "collect-support-bundle-for-kube-nodes_kube-master.tar.gz" and "collect-support-bundle-for-kube-nodes_kube-worker.tar.gz" ) and one for `collect-support-bundle-for-vm-external-node` ( "collect-support-bundle-for-vm-external-node_vm.tar.gz") showing in `/root/test` folder.

Author: updated

> showing in `/root/test` folder. Run the `tar xvf TARBALL_NAME` command and you will get the

Reviewer 2: ```suggestion
showing in `/root/test` folder. Run the `tar xvf $TARBALL_NAME` command and you will get the
```

Author: updated

# Support Bundle User Guide

## What is Support Bundle

Antrea supports collecting support bundle tarballs, which include the information
from Antrea Controller and Antrea Agent. It is useful when troubleshooting the issues
in Kubernetes clusters and external Nodes.

**Be aware that the generated support bundle includes a lot of information,
including logs, so please review the contents of the directory before sharing
it on Github and ensure that you do not share anything sensitive.**

> There are two ways of generating support bundles. Firstly, You can run `antctlsupportbundle`

Reviewer 2: ```suggestion
There are two ways of generating support bundles. Firstly, You can run `antctl supportbundle`
```

to designate this feature. The details will be provided in section [Use case example](#use-case-example).

## Table of Contents

<!-- toc -->
- [Prerequisites](#prerequisites)
- [Use case example](#use-case-example)
  - [Running antctl commands](#running-antctl-commands)
  - [Applying SupportBundleCollection CRD](#applying-supportbundlecollection-crd)
- [List of collected items](#list-of-collected-items)
- [Limitations](#limitations)
<!-- /toc -->

## Prerequisites

`SupportBundleCollection` is introduced in Antrea v1.10.0 as an alpha feature. The feature gate
must be enabled in both antrea-controller and antrea-agent configurations. If you plan
to collect support bundle on a specific external Node, you should enable it in the configuration
on the external Node as well.

```yaml
  antrea-agent.conf: |
    featureGates:
    # Enable collecting support bundle files with SupportBundleCollection CRD.
    SupportBundleCollection: true
```

```yaml
  antrea-controller.conf: |
    featureGates:
    # Enable collecting support bundle files with SupportBundleCollection CRD.
    SupportBundleCollection: true
```

> ## Use case example

Reviewer 2: Is it possible to introduce a section like "The SupportBundleCollection resource" to introduce the CRD before "Use case example"? Then we could guide users how to configure the fields in a CR, like the meaning and limitations of fields.

Author: Added. Personally I don't think we need to do the rewriting to explain the fields in the CRD one by one. Most of them are self-explaintory. Two examples in the `use case example` section are enough to illustrate all its functionalities.

Reviewer 2: Better to add a link to refer to the supportbundlecollection yaml file to let the users know how to create the resource

### Running antctl commands

Please refer to the corresponding [antctl user guide section](antctl.md#collecting-support-information).
Note: `antctl supportbundle` only works on collecting the support bundles from Antrea Controller and
Antrea Agent which is running on K8s Node, it does not work for the Agent on an external Node.

### Applying SupportBundleCollection CRD

> In this section, we will create two SupportBundleCollections for Antrea Nodes and ExternalNodes.

Reviewer 2: ```suggestion
In this section, we will create two SupportBundleCollections for K8s Nodes and ExternalNodes.
```

Author: updated

## Table of Contents

<!-- toc -->
- [Prerequisites](#prerequisites)
- [The SupportBundleCollection CRD](#the-supportbundlecollection-crd)
- [Use case example](#use-case-example)
  - [Running antctl commands](#running-antctl-commands)
  - [Applying SupportBundleCollection CRD](#applying-supportbundlecollection-crd)
- [List of collected items](#list-of-collected-items)
- [Limitations](#limitations)
<!-- /toc -->

## Prerequisites

`SupportBundleCollection` is introduced in Antrea v1.10.0 as an alpha feature. The feature gate
must be enabled in both antrea-controller and antrea-agent configurations. If you plan
to collect support bundle on a specific external Node, you should enable it in the configuration
on the external Node as well.

```yaml
  antrea-agent.conf: |
    featureGates:
    # Enable collecting support bundle files with SupportBundleCollection CRD.
    SupportBundleCollection: true
```

```yaml
  antrea-controller.conf: |
    featureGates:
    # Enable collecting support bundle files with SupportBundleCollection CRD.
    SupportBundleCollection: true
```

## The SupportBundleCollection CRD

> Collecting support bundle using `antctl` is limited to collecting local Pod

Reviewer 2: Remove the first sentence about antctl limitations.

or whole cluster information. SupportBundleCollection CR is introduced to supplement
the `antctl` command with three reinforcements:

> 1. Allow users to choose Nodes or ExternalNodes in batch.

Reviewer 2: ```suggestion
1. Allow users to collect support bundle files on ExternalNodes.
```

to designate this feature. The details will be provided in section [Use case example](#use-case-example).

## Table of Contents

<!-- toc -->
- [Prerequisites](#prerequisites)
- [The SupportBundleCollection CRD](#the-supportbundlecollection-crd)
- [Use case example](#use-case-example)
  - [Running antctl commands](#running-antctl-commands)
  - [Applying SupportBundleCollection CRD](#applying-supportbundlecollection-crd)
- [List of collected items](#list-of-collected-items)
- [Limitations](#limitations)
<!-- /toc -->

## Prerequisites

`SupportBundleCollection` is introduced in Antrea v1.10.0 as an alpha feature. The feature gate
must be enabled in both antrea-controller and antrea-agent configurations. If you plan
to collect support bundle on a specific external Node, you should enable it in the configuration
on the external Node as well.

```yaml
  antrea-agent.conf: |
    featureGates:
    # Enable collecting support bundle files with SupportBundleCollection CRD.
    SupportBundleCollection: true
```

```yaml
  antrea-controller.conf: |
    featureGates:
    # Enable collecting support bundle files with SupportBundleCollection CRD.
    SupportBundleCollection: true
```

> A single namespace (e.g., default) is created for saving the secrets which are

Reviewer: ```suggestion
A single Namespace (e.g., default) is created for saving the secrets which are
```

Reviewer 2: done.

used to access the support bundle file server, and the permission to read secrets

> in this namespace is given to antrea-controller by modifying and applying the

Reviewer: ditto, please check other occurrences.

Reviewer 2: sure, updated.

[RBAC file](/build/yamls/externalnode/support-bundle-collection-rbac.yml).

```yaml

> ...

Reviewer: Remove this line?

Reviewer 2: done.

kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: antrea-read-secrets
  namespace: default  # Change the Namespace to where the Secret for file server's authentication credential is created.
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: antrea-secret-reader
subjects:
  - kind: ServiceAccount
    name: antrea-controller
    namespace: kube-system
```

## The SupportBundleCollection CRD

> SupportBundleCollection CR is introduced to supplement the `antctl` command with
> three reinforcements:

Reviewer: ```suggestion
SupportBundleCollection CRD is introduced to supplement the `antctl` command with
three additional features:
```

Reviewer 2: updated.

1. Allow users to collect support bundle files on ExternalNodes.
2. Upload all the support bundle files into a user-provided SFTP Server.
3. Support tracking status of a SupportBundleCollection CR.

Details are provided in the next section.

## Use case example

### Running antctl commands

Please refer to the corresponding [antctl user guide section](antctl.md#collecting-support-information).
Note: `antctl supportbundle` only works on collecting the support bundles from Antrea Controller and
Antrea Agent which is running on K8s Node, it does not work for the Agent on an external Node.

### Applying SupportBundleCollection CRD

In this section, we will create two SupportBundleCollections for Kubernetes Nodes and ExternalNodes.

> Assume we have a cluster with Nodes named "kube-worker" and "kube-master". In addition, we have

Reviewer 3: We should not use word "master". Consider "worker1", "worker2".

> set up an ExternalNode as `vm` in the `vm-ns` Namespace by following the instruction

Reviewer 3: Maybe `vm1`

in [Install ExternalNode on VM](external-node.md#install-antrea-agent-on-vm).

A Secret needs to be created in advance with the username and password of the SFTP Server.

> The Secret will be used in `authSecret` in the following YAML examples.

Reviewer 3: will be referred as

```bash
kubectl create secret generic support-bundle-secret  --from-literal=username='your-sftp-username'  --from-literal=password='your-sftp-password'
```

> Then we could apply the following YAML files. The first one is to collect support bundle

Reviewer 3: could -> can

on Kubernetes Nodes "kube-worker" and "kube-master". The second one is to collect support
bundle on ExternalNode "vm".

```bash
cat << EOF | kubectl apply -f -
kind: SupportBundleCollection
metadata:
  name: collect-support-bundle-for-kube-nodes
spec:
  nodes:
    nodeNames: # Empty list means all the Nodes
      - kube-master
      - kube-worker
    matchLabels:

>       app: nginx   # Collect support bundle files from the Nodes which have been labeled as "app: nginx".

Reviewer 3: Should we comment it out? Or you should add description about Nodes with the nginx label.

  expirationMinutes: 10 # expirationMinutes is the requested duration of validity of the SupportBundleCollection. A SupportBundleCollection will be marked as Failed if it does not finish before expiration.

>   sinceTime: 2h # Collect the logs in the latest 2 hours. Collect all Agent logs if it is not set.

Reviewer 3: Remove "Agent".

all available logs if the time is not set

  fileServer:
    url: sftp://yourtestdomain.com:22/root/test
  authentication:
    authType: "BasicAuthentication"
    authSecret:
      name: support-bundle-secret

>       namespace: default # antrea-controller has been given the permission to read secrets in "default" namespace. 

Reviewer 3: has been -> must be

Secrets
Namespace

EOF
```

```bash
cat << EOF | kubectl apply -f -
apiVersion: crd.antrea.io/v1alpha1
kind: SupportBundleCollection
metadata:

>   name: collect-support-bundle-for-vm-external-node

Reviewer 3: Remove "external"?

spec:
  externalNodes:

>     nodeNames: # Empty list means all the ExternalNodes in the Namespace

Reviewer 3: It may be ok to use `ExternalNode` in such cases, which can refer to the ExternalNode CR

      - vm
    nodeSelector:
      matchLabels:

>         app: db   # Collect support bundle files from the ExternalNodes which have been labeled as "app: db" in the Namespace

Reviewer 3: The same comment on label.

    namespace: vm-ns # namespace is mandatory when collecting files from ExternalNodes.
  fileServer:
    url: yourtestdomain.com:22/root/test # Scheme sftp can be omitted
  authentication:
    authType: "BasicAuthentication"
    authSecret:
      name: support-bundle-secret
      namespace: default # antrea-controller has been given the permission to read secrets in "default" namespace.
EOF
```

> More details about the supported fields in a "SupportBundleCollection" CR, please

Reviewer 3: For more information

refer to the [CRD definition](/build/charts/antrea/crds/supportbundlecollection.yaml)

You can check the status of `SupportBundleCollection` by running
`kubectl get supportbundlecollections $Name-of-SupportBundleCollection -ojson`. The following example

> shows a successful creation of a `SupportBundleCollection`. `desiredNodes` shows the number of

Reviewer 3: creation -> realization?

a `SupportBundleCollection` -> `SupportBundleCollection`

> Nodes/ExternalNodes is expected in this request, while `collectedNodes` shows the number of

Reviewer 3: Here too

> Nodes/ExternalNodes which has already uploaded bundle files to the target file server.`collectedNodes`

Reviewer 3: Add a space after "."

and `desiredNodes` should match the number of Nodes/ExternalNodes you want to collect support

> bundle if collection is running successfully.

Reviewer 3: the collection

is running -> succeeded (or "completed successfully")

> If following two conditions are presented, it means a bundle creation is succeeded,

Reviewer 3: the following two

creation -> collection

1. "Completed" is true
2. "CollectionFailure" is false.

> If any expected Node/ExternalNode is failed to upload the bundle files in the required time, the

Reviewer 3: is failed -> failed

> "CollectionFailure" condition is marked as true.

Reviewer 3: will be set to true

```bash
$ kubectl get supportbundlecollections collect-support-bundle-for-kube-nodes -ojson

...
 "status": {
        "collectedNodes": 1,
        "conditions": [
            {
                "lastTransitionTime": "2022-12-08T06:49:35Z",
                "status": "True",
                "type": "Started"
            },
            {
                "lastTransitionTime": "2022-12-08T06:49:41Z",
                "status": "True",
                "type": "BundleCollected"
            },
            {
                "lastTransitionTime": "2022-12-08T06:49:35Z",
                "status": "False",
                "type": "CollectionFailure"
            },
            {
                "lastTransitionTime": "2022-12-08T06:49:41Z",
                "status": "True",
                "type": "Completed"
            }
        ],
        "desiredNodes": 1
    }
```

> The output contains three tarballs. To access these files, please SSH into the ftp server

Reviewer 3: output -> collection

Reviewer 3: contains -> should include

Reviewer 3: SFTP

> `yourtestdomain.com` with port 22 or scp them to any other storages you used in CR. There will be

Reviewer 3: Why you need to SSH to a SFTP server, rather than use "SFTP"? Maybe just say "you can download the files from the SFTP server"

> two tarballs for collect-support-bundle-for-kube-nodes ("collect-support-bundle-for-kube-nodes_kube-master.tar.gz" and "collect-support-bundle-for-kube-nodes_kube-worker.tar.gz") and

Reviewer: It's too long, you need wrap the line.

> two tarballs for collect-support-bundle-for-kube-nodes ("collect-support-bundle-for-kube-nodes_kube-master.tar.gz" and "collect-support-bundle-for-kube-nodes_kube-worker.tar.gz") and

Reviewer 3: Consider using '' to wrap `collect-support-bundle-for-kube-nodes`

> two tarballs for collect-support-bundle-for-kube-nodes ("collect-support-bundle-for-kube-nodes_kube-master.tar.gz" and "collect-support-bundle-for-kube-nodes_kube-worker.tar.gz") and
> one for collect-support-bundle-for-vm-external-node ( "collect-support-bundle-for-vm-external-node_vm.tar.gz")

Reviewer: Maybe like below?
```
```
```suggestion
two tarballs for collect-support-bundle-for-kube-nodes including "collect-support-bundle-for-kube-nodes_kube-master.tar.gz"
and "collect-support-bundle-for-kube-nodes_kube-worker.tar.gz", and one for collect-support-bundle-for-vm-external-node which is "collect-support-bundle-for-vm-external-node_vm.tar.gz"
```

Reviewer 2: updated accordingly.

> showing in `/root/test` folder. Run the `tar xvf $TARBALL_NAME` command, and you will get the

Reviewer 3: Remove "showing"

in the `/root/test` folder

Reviewer 3: "Run the `tar xvf $TARBALL_NAME` command to extract the files from the tarballs".

> files, each with different collect items.

Reviewer 3: I guess remove ", each with different collect items"

## List of collected items

> Depending on the methods you use to collect support bundle, the items you collect may differ.

Reviewer 3: to collect the

Reviewer 3: the contents in the bundle may differ

> Here is the table that shows such differences.

Reviewer: ```suggestion
Here is the table that shows the differences.
```

Reviewer 3: The following table shows the differences

> We use `A`,`C` and `O` to represent "running `antctl supportbundle` in the Antrea Agent",

Reviewer 3: This is hard to understand. How about `agent`, `controller`, `outside`

> We use `A`,`C` and `O` to represent "running `antctl supportbundle` in the Antrea Agent",
> "running `antctl supportbundle` in the Antrea Controller" and
> "running `antctl supportbundle` out-of-cluster". Also, We use `SBC-N` and `SBC-E` to represent

Reviewer: ```suggestion
We use `A`,`C` and `O` to represent "running `antctl supportbundle` in the Antrea Agent",
"running `antctl supportbundle` in the Antrea Controller" and
"running `antctl supportbundle` out-of-cluster" respectively. Also, we use `SBC-N` and `SBC-E` to represent
```

Reviewer 2: updated.

> "running `antctl supportbundle` out-of-cluster". Also, We use `SBC-N` and `SBC-E` to represent

Reviewer 3: Add "respectively"

Reviewer 3: Also -> And

Reviewer 3: SBC-N -> Node
SBC-E -> ExternalNode

> "apply SupportBundleCollection CRD to create support bundle on Nodes" and

Reviewer 3: "create SupportBundleCollection CR for Nodes"

"apply SupportBundleCollection CRD to create support bundle on ExternalNodes".

| Collected Item             | Supported Collecting Method | Explanation                                                                                                                                                                                                                                                           |
| -------------------------- | --------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | Antrea Agent Log           | A, O, SBC-N, SBC-E          | Log file of Antrea Agent                                                                                                                                                                                                                                              |

Reviewer 3: Log file of Antrea Agent -> Antrea Agent log files

| Antrea Controller Log      | C, O                        | Log file of Antrea Controller                                                                                                                                                                                                                                         |

> | iptables(Linux Only)       | A, O, SBC-N, SBC-E          | Result of running `ip6tables-save` and `iptable-save` with counters                                                                                                                                                                                                   |

Reviewer: ```suggestion
| iptables (Linux Only)       | A, O, SBC-N, SBC-E          | Result of running `ip6tables-save` and `iptable-save` with counters                                                                                                                                                                                                   |
```

Reviewer 2: done.

> | iptables(Linux Only)       | A, O, SBC-N, SBC-E          | Result of running `ip6tables-save` and `iptable-save` with counters                                                                                                                                                                                                   |

Reviewer 3: Probably change all "Result" to "Output".

> | OVS Ports                  | A, O, SBC-N, SBC-E          | Result of running `ovs-ofctl dump-ports-desc`                                                                                                                                                                                                                         |

Reviewer 3: Remove "running"

> | NetworkPolicy Resources    | A, C, O, SBC-N, SBC-E       | YAML output of running `antctl get appliedtogroups`, `antctl get addressgroups`                                                                                                                                                                                       |

Reviewer 3: of `antctl get appliedtogroups` and `antctl get addressgroups` commands

| Heap Pprof                 | A, C, O, SBC-N, SBC-E       | Result of running [`pprof.WriteHeapProfile`](https://pkg.go.dev/runtime/pprof#WriteHeapProfile)                                                                                                                                                                       |

> | HNSResources(Windows Only) | A, O, SBC-N, SBC-E          | Result of running `Get-HnsNetwork` and `Get-HNSEndpoint`                                                                                                                                                                                                              |

Reviewer: ditto

Reviewer: Question:
is this correct `Get-HnsNetwork` and `Get-HNSEndpoint` ?
one is Hns*, another is HNS*

Reviewer 2: Actually, Windows is insensitive on the character upper case or lower case, so both work well in Windows powershell. The standard command should be "HNS", updated.

> | HNSResources(Windows Only) | A, O, SBC-N, SBC-E          | Result of running `Get-HnsNetwork` and `Get-HNSEndpoint`                                                                                                                                                                                                              |

Reviewer 3: of `Get-HnsNetwork` and `Get-HNSEndpoint` commands

> | Antrea Agent Info          | A, O, SBC-N, SBC-E          | YAML output of running `antctl get agentinfo`                                                                                                                                                                                                                         |

Reviewer 3: Remove running

> | Antrea Controller Info     | C, O                        | YAML output of running `antctl get controllerinfo`                                                                                                                                                                                                                    |

Reviewer 3: Remove running

> | IP Address info            | A, O, SBC-N, SBC-E          | Result of running `ip address` in Linux or `ipconfig /all` in Windows                                                                                                                                                                                                 |

Reviewer 3: "Info" to be consistent with others

> | IP Route info              | A, O, SBC-N, SBC-E          | Result of running `ip route` in Linux or `route print` in Windows                                                                                                                                                                                                     |

Reviewer 3: Ditto

Reviewer 3: of `ip address` command on Linux or `ipconfig /all` command on Windows

> | IP Link info               | A, O, SBC-N, SBC-E          | Result of running `ip link` in Linux or `Get-NetAdapter` in Windows                                                                                                                                                                                                   |

Reviewer 3: Ditto.

> | Cluster Information        | O                           | Concatenation of information, including: 1. all Pods, Deployments, Replicasets and Daemonsets in all namespaces with any resourceVersion. 2. all nodes with any resourceVersion. 3. all ConfigMaps in all namespaces with any resourceVersion and label `app=antrea`. |

Reviewer: ```suggestion
| Cluster Information        | O                           | Concatenation of information, including: 1. all Pods, Deployments, Replicasets and Daemonsets in all Namespaces with any resourceVersion. 2. all Nodes with any resourceVersion. 3. all ConfigMaps in all Namespaces with any resourceVersion and label `app=antrea`. |
```

Reviewer 2: done.

> | Cluster Information        | O                           | Concatenation of information, including: 1. all Pods, Deployments, Replicasets and Daemonsets in all namespaces with any resourceVersion. 2. all nodes with any resourceVersion. 3. all ConfigMaps in all namespaces with any resourceVersion and label `app=antrea`. |

Reviewer 3: @wenyingd : we collect all Pods, etc. in the cluster? Why? I feel we should not do this. Not good to collect all workload information from customer clusters either.

Reviewer 3: I feel this is wrong. But let us discuss and handle that separately. @tnqn @edwardbadboy

# Support Bundle User Guide

## What is Support Bundle

Antrea supports collecting support bundle tarballs, which include the information
from Antrea Controller and Antrea Agent. The collected information can help
debugging issues in the Kubernetes cluster.

**Be aware that the generated support bundle includes a lot of information,
including logs, so please review the contents before sharing it on Github
and ensure that you do not share any sensitive information.**

There are two ways of generating support bundles. Firstly, you can run `antctl supportbundle`
directly in the Antrea Agent Pod, Antrea Controller Pod, or on a host with a
`kubeconfig` file for the target the cluster. Secondly, you can also apply

> `SupportBundleCollection` CRs to create support bundles for Kubernetes Nodes

Reviewer 3: I saw you use K8s in other places. Then probably change this one too

Reviewer 2: updated.

roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: antrea-secret-reader
subjects:
  - kind: ServiceAccount
    name: antrea-controller
    namespace: kube-system
```

## The SupportBundleCollection CRD

SupportBundleCollection CRD is introduced to supplement the `antctl` command
with three additional features:

1. Allow users to collect support bundle files on external Nodes.
2. Upload all the support bundle files into a user-provided SFTP Server.
3. Support tracking status of a SupportBundleCollection CR.

Details are provided in the next section.

## Usage examples

### Running antctl commands

Please refer to the [antctl user guide section](antctl.md#collecting-support-information).
Note: `antctl supportbundle` only works on collecting the support bundles from
Antrea Controller and Antrea Agent that is running on a K8s Node, but it does
not work for the Agent on an external Node.

### Applying SupportBundleCollection CR

In this section, we will create two SupportBundleCollection CRs for K8s Nodes
and external Nodes. Assume we have a cluster with Nodes named "worker1" and
"worker2". In addition, we have set up two external Nodes named "vm1" and "vm2"

> in the "vm-ns" Namespace by following the instruction in [Install ExternalNode on VM](external-node.md#install-antrea-agent-on-vm).

Reviewer 3: `in [Install ExternalNode on VM]` -> `of the [VM installation guide]`

Reviewer 2: updated.

A Secret needs to be created in advance with the username and password of the SFTP
Server. The Secret will be referred as `authSecret` in the following YAML examples.

```bash
kubectl create secret generic support-bundle-secret  --from-literal=username='your-sftp-username'  --from-literal=password='your-sftp-password'
```

Then we can apply the following YAML files. The first one is to collect support

> bundle on Kubernetes Nodes "worker1" and "worker2", and "worker2" is labeled

Reviewer 3: K8s

Reviewer 2: updated.

> as "app: nginx". The second one is to collect support bundle on ExternalNode "vm1"

Reviewer 3: I feel "external Node" works better here.

Reviewer 2: Updated as `to collect support bundle on external Nodes "vm1" and "vm2" in Namespace "vm-ns", and ExternalNode "vm2" is labeled with ...`

> as "app: nginx". The second one is to collect support bundle on ExternalNode "vm1"

Reviewer 2: updated.

> and "vm2" in Namespace "vm-ns", and "vm2" is labeled as "app: db".

Reviewer 2: updated.

```bash
cat << EOF | kubectl apply -f -
kind: SupportBundleCollection
metadata:
  name: support-bundle-for-nodes
spec:
  nodes:
    nodeNames: # Empty list means all the Nodes
      - worker1
    matchLabels:
      app: nginx
  expirationMinutes: 10 # expirationMinutes is the requested duration of validity of the SupportBundleCollection. A SupportBundleCollection will be marked as Failed if it does not finish before expiration.
  sinceTime: 2h # Collect the logs in the latest 2 hours. Collect all available logs if the time is not set.
  fileServer:
    url: sftp://yourtestdomain.com:22/root/test
  authentication:
    authType: "BasicAuthentication"
    authSecret:
      name: support-bundle-secret
      namespace: default # antrea-controller must be given the permission to read Secrets in "default" Namespace. 
EOF
```

```bash
cat << EOF | kubectl apply -f -
apiVersion: crd.antrea.io/v1alpha1
kind: SupportBundleCollection
metadata:
  name: support-bundle-for-vms
spec:
  externalNodes:

>     nodeNames: # Empty list means all ExternalNode in the Namespace

Reviewer 2: updated.

      - vm1
    nodeSelector:
      matchLabels:
        app: db
    namespace: vm-ns # namespace is mandatory when collecting files from external Nodes.
  fileServer:
    url: yourtestdomain.com:22/root/test # Scheme sftp can be omitted
  authentication:
    authType: "BasicAuthentication"
    authSecret:
      name: support-bundle-secret
      namespace: default # antrea-controller must be given the permission to read Secrets in "default" Namespace.
EOF
```

For more information about the supported fields in a "SupportBundleCollection"
CR, please refer to the [CRD definition](/build/charts/antrea/crds/supportbundlecollection.yaml)

You can check the status of `SupportBundleCollection` by running
`kubectl get supportbundlecollections $Name-of-SupportBundleCollection -ojson`.
The following example shows a successful realization of `SupportBundleCollection`.

> `desiredNodes` shows the number of Node/ExternalNode is expected in this request,

Reviewer 3: the expected number of Nodes/ExternalNodes to collect information with this request

> `desiredNodes` shows the number of Node/ExternalNode is expected in this request,

Reviewer 2: updated.

> while `collectedNodes` shows the number of Node/ExternalNode which has already

Reviewer 3: has -> have

> while `collectedNodes` shows the number of Node/ExternalNode which has already

Reviewer 2: updated.

uploaded bundle files to the target file server. `collectedNodes` and `desiredNodes`

> should match the number of Node/ExternalNode you want to collect support bundle

Reviewer 2: updated.

> if the collection is succeeded.

Reviewer 3: The sentence is a little long. Probably reads better, if we move "if..." to the beginning:
> If the collection successfully completes, `collectedNodes` and `desiredNodes` should..."

> if the collection is succeeded.

Reviewer 2: updated.

If the following two conditions are presented, it means a bundle collection is

> succeeded,

Reviewer 2: updated.

1. "Completed" is true
2. "CollectionFailure" is false.

If any expected Node/ExternalNode failed to upload the bundle files in the
required time, the "CollectionFailure" condition will be set to true.

```bash
$ kubectl get supportbundlecollections support-bundle-for-nodes -ojson

...
 "status": {
        "collectedNodes": 1,
        "conditions": [
            {
                "lastTransitionTime": "2022-12-08T06:49:35Z",
                "status": "True",
                "type": "Started"
            },
            {
                "lastTransitionTime": "2022-12-08T06:49:41Z",
                "status": "True",
                "type": "BundleCollected"
            },
            {
                "lastTransitionTime": "2022-12-08T06:49:35Z",
                "status": "False",
                "type": "CollectionFailure"
            },
            {
                "lastTransitionTime": "2022-12-08T06:49:41Z",
                "status": "True",
                "type": "Completed"
            }
        ],
        "desiredNodes": 1
    }
```

> The collection should include three tarballs. To access these files, you can

Reviewer 2: updated.

download the files from the SFTP server `yourtestdomain.com`. There will be two

> tarballs for `support-bundle-for-nodes` including

Reviewer 2: updated.

"support-bundle-for-nodes_worker1.tar.gz" and "support-bundle-for-nodes_worker2.tar.gz",

> and two for `support-bundle-for-vms` including "support-bundle-for-vms_vm1.tar.gz"

Reviewer 3: Ditto

Reviewer 2: done.

> and "support-bundle-for-vms_vm2.tar.gz" in the `/root/test` folder. Run the

Reviewer 3: Maybe add a "," before "in the.."

Reviewer 2: updated.

`tar xvf $TARBALL_NAME` command to extract the files from the tarballs.

## List of collected items

Depending on the methods you use to collect the support bundle, the contents in
the bundle may differ. The following table shows the differences.

We use `agent`,`controller` and `outside` to represent running command

> `antctl supportbundle` in Antrea Agent, Antrea Controller and out-of-cluster

Reviewer 3: I feel changing "and" to "," reads better.

Reviewer 2: updated.

respectively. Also, we use `Node` and `ExternalNode` to represent
"create SupportBundleCollection CR for Nodes" and "create SupportBundleCollection
CR for external Nodes".

| Collected Item              | Supported Collecting Method                     | Explanation                                                                                                                                                                                                                                                           |
|-----------------------------|-------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | Antrea Agent Log            | agent, outside, Node, ExternalNode              | Antrea Agent log files                                                                                                                                                                                                                                                |

Reviewer 3: It may look better to wrap agent, etc. with ``: `agent`, `outside`, ...

Reviewer 2: updated.

# Support Bundle User Guide

## What is Support Bundle

Antrea supports collecting support bundle tarballs, which include the information
from Antrea Controller and Antrea Agent. The collected information can help
debugging issues in the Kubernetes cluster.

**Be aware that the generated support bundle includes a lot of information,
including logs, so please review the contents before sharing it on Github
and ensure that you do not share any sensitive information.**

There are two ways of generating support bundles. Firstly, you can run `antctl supportbundle`
directly in the Antrea Agent Pod, Antrea Controller Pod, or on a host with a

> `kubeconfig` file for the target the cluster. Secondly, you can also apply

Reviewer 4: ```suggestion
`kubeconfig` file for the target cluster. Secondly, you can also apply
```

Reviewer 2: done.

  - [Applying SupportBundleCollection CR](#applying-supportbundlecollection-cr)
- [List of collected items](#list-of-collected-items)
- [Limitations](#limitations)
<!-- /toc -->

## Prerequisites

The `antctl supportbundle` command is supported in Antrea since version 0.7.0.

The `SupportBundleCollection` CRD is introduced in Antrea v1.10.0 as an alpha
feature. The feature gate must be enabled in both antrea-controller and
antrea-agent configurations. If you plan to collect support bundle on an external
Node, you should enable it in the configuration on the external Node as well.

```yaml
  antrea-agent.conf: |
    featureGates:
    # Enable collecting support bundle files with SupportBundleCollection CRD.
    SupportBundleCollection: true
```

```yaml
  antrea-controller.conf: |
    featureGates:
    # Enable collecting support bundle files with SupportBundleCollection CRD.
    SupportBundleCollection: true
```

A single Namespace (e.g., default) is created for saving the Secrets that are
used to access the support bundle file server, and the permission to read Secrets
in this Namespace is given to antrea-controller by modifying and applying the

> [RBAC file](/build/yamls/externalnode/support-bundle-collection-rbac.yml).

Reviewer 4: don't use absolute path which will be a broken link in antrea.io, though I don't know why the verification stopped alarming this.

Reviewer 2: updated.

```yaml
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: antrea-read-secrets
  namespace: default  # Change the Namespace to where the Secret for file server's authentication credential is created.
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: antrea-secret-reader
subjects:
  - kind: ServiceAccount
    name: antrea-controller
    namespace: kube-system
```

## The SupportBundleCollection CRD

SupportBundleCollection CRD is introduced to supplement the `antctl` command
with three additional features:

1. Allow users to collect support bundle files on external Nodes.
2. Upload all the support bundle files into a user-provided SFTP Server.
3. Support tracking status of a SupportBundleCollection CR.

> Details are provided in the next section.

Reviewer 4: This seems too verbose if it just points to the next section.

Reviewer 2: removed this sentence.

## Usage examples

### Running antctl commands

Please refer to the [antctl user guide section](antctl.md#collecting-support-information).
Note: `antctl supportbundle` only works on collecting the support bundles from
Antrea Controller and Antrea Agent that is running on a K8s Node, but it does
not work for the Agent on an external Node.

### Applying SupportBundleCollection CR

In this section, we will create two SupportBundleCollection CRs for K8s Nodes
and external Nodes. Assume we have a cluster with Nodes named "worker1" and
"worker2". In addition, we have set up two external Nodes named "vm1" and "vm2"
in the "vm-ns" Namespace by following the instruction of the [VM installation guide](external-node.md#install-antrea-agent-on-vm).

A Secret needs to be created in advance with the username and password of the SFTP
Server. The Secret will be referred as `authSecret` in the following YAML examples.

```bash
kubectl create secret generic support-bundle-secret  --from-literal=username='your-sftp-username'  --from-literal=password='your-sftp-password'
```

Then we can apply the following YAML files. The first one is to collect support

> bundle on K8s Nodes "worker1" and "worker2", and "worker2" is labeled with

Reviewer 4: I feel it should have a statement about user can specify nodes by their names or by matching their labels first, then give this example and say it selects worker1 by specifying its name, and selects worker2 by matching its label.

Reviewer 2: Add a statement in the first paragraph.

> "app: nginx". The second one is to collect support bundle on external Nodes

Reviewer 4: this is not a good example of node label, please use some better examples.

Reviewer 2: Use labels "role: workers" and "role: vms" in the example.

  authentication:
    authType: "BasicAuthentication"
    authSecret:
      name: support-bundle-secret
      namespace: default # antrea-controller must be given the permission to read Secrets in "default" Namespace. 
EOF
```

```bash
cat << EOF | kubectl apply -f -
apiVersion: crd.antrea.io/v1alpha1
kind: SupportBundleCollection
metadata:
  name: support-bundle-for-vms
spec:
  externalNodes:
    nodeNames: # Empty list means all ExternalNodes in the Namespace
      - vm1
    nodeSelector:
      matchLabels:
        app: db
    namespace: vm-ns # namespace is mandatory when collecting files from external Nodes.
  fileServer:
    url: yourtestdomain.com:22/root/test # Scheme sftp can be omitted
  authentication:
    authType: "BasicAuthentication"
    authSecret:
      name: support-bundle-secret
      namespace: default # antrea-controller must be given the permission to read Secrets in "default" Namespace.
EOF
```

For more information about the supported fields in a "SupportBundleCollection"
CR, please refer to the [CRD definition](/build/charts/antrea/crds/supportbundlecollection.yaml)

You can check the status of `SupportBundleCollection` by running

> `kubectl get supportbundlecollections $Name-of-SupportBundleCollection -ojson`.

Reviewer 4: it's not common to given example in this format "$Name-of-SupportBundleCollection", perhaps just `kubectl get supportbundlecollections [NAME] -ojson`

Reviewer 2: updated accordingly.

The following example shows a successful realization of `SupportBundleCollection`.

> `desiredNodes` shows the number of Nodes/ExternalNodes to collect information

Reviewer 3: "the expected number of"

Reading it again, guess we should remove "information".

| IP Address Info             | `agent`, `outside`, `Node`, `ExternalNode`               | Output of `ip address` command on Linux or `ipconfig /all` command on Windows                                                                                                                                                                                         |
| IP Route Info               | `agent`, `outside`, `Node`, `ExternalNode`               | Output of `ip route` on Linux or `route print` on Windows                                                                                                                                                                                                             |
| IP Link Info                | `agent`, `outside`, `Node`, `ExternalNode`               | Output of `ip link` on Linux or `Get-NetAdapter` on Windows                                                                                                                                                                                                           |

> | Cluster Information         | `outside`                                                | Concatenation of information, including: 1. all Pods, Deployments, Replicasets and Daemonsets in all Namespaces with any resourceVersion. 2. all Nodes with any resourceVersion. 3. all ConfigMaps in all Namespaces with any resourceVersion and label `app=antrea`. |

Reviewer 3: "Concatenation of information" -  I do not know this word well. If you are not sure either, consider: "Dump of resources in the cluster".

names or by matching their labels in a SupportBundleCollection CR.

Assume we have a cluster with Nodes named "worker1" and "worker2". In addition,
we have set up two external Nodes named "vm1" and "vm2" in the "vm-ns" Namespace
by following the instruction of the [VM installation guide](external-node.md#install-antrea-agent-on-vm).

A Secret needs to be created in advance with the username and password of the SFTP
Server. The Secret will be referred as `authSecret` in the following YAML examples.

```bash
kubectl create secret generic support-bundle-secret  --from-literal=username='your-sftp-username'  --from-literal=password='your-sftp-password'
```

Then we can apply the following YAML files. The first one is to collect support
bundle on K8s Nodes "worker1" and "worker2": "worker1" is specified by the name,
and "worker2" is specified by matching label "role: workers". The second one is to
collect support bundle on external Nodes "vm1" and "vm2" in Namespace "vm-ns":
"vm1" is specified by the name, and "vm2" is specified by matching label "role: vms".

```bash
cat << EOF | kubectl apply -f -
kind: SupportBundleCollection
metadata:
  name: support-bundle-for-nodes
spec:
  nodes:

>     nodeNames: # Empty list means all the Nodes

Reviewer 4: When nodeNames is empty but matchLabels is set, does the `nodeName` also means all Nodes? I guess not. Then it's perhaps to move the comment to `nodes:`:
```
nodes:  # All Nodes will be selected if both nodeNames and matchLabels are empty.
```

Reviewer 2: updated.

      - worker1
    matchLabels:
      role: workers
  expirationMinutes: 10 # expirationMinutes is the requested duration of validity of the SupportBundleCollection. A SupportBundleCollection will be marked as Failed if it does not finish before expiration.
  sinceTime: 2h # Collect the logs in the latest 2 hours. Collect all available logs if the time is not set.
  fileServer:
    url: sftp://yourtestdomain.com:22/root/test
  authentication:
    authType: "BasicAuthentication"
    authSecret:
      name: support-bundle-secret
      namespace: default # antrea-controller must be given the permission to read Secrets in "default" Namespace. 
EOF
```

```bash
cat << EOF | kubectl apply -f -
apiVersion: crd.antrea.io/v1alpha1
kind: SupportBundleCollection
metadata:
  name: support-bundle-for-vms
spec:
  externalNodes:

>     nodeNames: # Empty list means all ExternalNodes in the Namespace

Reviewer 4: ditto

Reviewer 2: updated.
